### ISSUE TITLE
twisted.internet.endpoints.clientFromString should be ported to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7973]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7973|
|Type|enhancement|
|Created|2015-07-19 09:07:39Z|
|Branch|https://github.com/twisted/twisted/tree/clientFromString-py3-7973|

<details><summary>Searchable metadata</summary>

```
trac-id__7973 7973
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_clientFromString_py3_7973 branches/clientFromString-py3-7973
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1437296859827207 1437296859827207
changetime__1438400779914554 1438400779914554
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [45275]) Branching to clientFromString-py3-7973.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

Builders are looking acceptable. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @hawkowl|
|-|-|

This ticket talks about clientFromString but I saw that it also touched the serverFromString part.... but serverStringTests were not enabled.

Do we want to use some vague type like `C{str}`? I prefer to use either bytes or Unicode.

What do you think?

Thanks for working on this. The patch looks good.

Feel free to address my comments and merge!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

Hi Adi, that's because there's common functions, so I have to touch that function's docs that talk to serverFromString. I'll be porting that momentarily.

str is because it actually needs to be a native string -- bytes on Py2.7, Unicode on Py3.3+.

Will merge.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl set status to `closed`|
|-|-|

(In [45321]) Merge clientFromString-py3-7973: Port clientFromString to Py3

Author: hawkowl
Reviewer: adiroiban
Fixes: [#7973](https://github.com/twisted/twisted/issues/7973)
